keps/sig-windows/0000-20181221-windows-group-managed-service-accounts-for-container-identity.md

```diff
@@ -0,0 +1,225 @@
+
+When this app moves to production on containers, the team will need to coordinate to make sure the right GMSA is carried over to the container deployment.
+
+1. Active Directory domain administrator will:
+
+  - Join all Windows Kubernetes nodes to the Active Directory domain.
+  - Provision the service account and gives details to application admin.
+  - Assign access to a security group to control what machines can use this service account. This group should include all authorized Kubernetes nodes.
+
+2. A Kubernetes admin will create a namespace, configmap and service account:
+  - Create a unique Kubernetes [service account](https://kubernetes.io/docs/tasks/configure-pod-container/configure-service-account/#use-multiple-service-accounts) as an identity for the application in Kubernetes. No rights to the cluster are assigned.
+
+```
+apiVersion: v1
+kind: ServiceAccount
+metadata:
+  name: webserver
+```
+
+  - Run a Windows PowerShell script to verify the account exists, and capture enough data to uniquely identify the GMSA into a JSON file (credspec). This doesn't contain any passwords or crypto secrets.
+
+Example credspec.json
+```
+{
+  "CmsPlugins": [
+    "ActiveDirectory"
+  ],
+  "DomainJoinConfig": {
+    "DnsName": "contoso.com",
+    "Guid": "244818ae-87ca-4fcd-92ec-e79e5252348a",
+    "DnsTreeName": "contoso.com",
+    "NetBiosName": "DEMO",
+    "Sid": "S-1-5-21-2126729477-2524075714-3094792973",
+    "MachineAccountName": "WebApplication1"
+  },
+  "ActiveDirectoryConfig": {
+    "GroupManagedServiceAccounts": [
+      {
+        "Name": "WebApplication1",
+        "Scope": "DEMO"
+      },
+      {
+        "Name": "WebApplication1",
+        "Scope": "contoso.com"
+      }
+    ]
+  }
+}
+```
+
+  - Create a configmap in the namespace with the contents of the JSON - `kubectl create configmap webserverCredSpec --from-file=gmsa-cred-spec=/path/to/credspec.json`
+  - Add reference to the configmap in the service account - `kubectl patch serviceaccount/webserver -p '{\"windowsCredentialSpec\": \"webserverCredSpec\"}'`
+3. The Kubernetes admin gives a set of users rights to the namespace and service account
+4. Application admin deploys the app with `ServiceAccount: webserver` as part of the PodSpec
+5. The ServiceAccount admission controller validates this user has access to the service account, and sets `PodSpec.SecurityContext.WindowsGMSACredentialSpec` to the contents of the credspec configmap
+6. The Windows CRI implementation copies the credentialSpec to the [OCI windows.CredentialSpec](https://github.com/opencontainers/runtime-spec/blob/master/config-windows.md#credential-spec) field
```

Comment:
The configmap would belong to a namespace that would be authorized for the user.  This would prevent other users of the cluster from being able to leverage a GMSA that they are not allowed to access.

Reply:
> The configmap would belong to a namespace that would be authorized for the user. This would prevent other users of the cluster from being able to leverage a GMSA that they are not allowed to access.

I'm not following that. The data in the configmap is not a secret, correct? And a user could forego the serviceaccount->configmap reference and just set the data directly in their podspec, right?